### PR TITLE
Elite Strike Team Location Track Assister 9000

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

Implements a new feature for elite strike teams (Nuke teams and ERTs) - Team tracking.
All syndicate hardsuits and ERT hardsuits are now equipped with a tracking beacon and a tracking HUD.
When the HUD is enabled it will point towards any beacons active on that frequency.
This means that any of your teammates will have a big marker pointing towards them.

 - Different nuke / ERT teams receive different codes
 - ERTs and Nuclear Operatives operate on a different base frequency so cannot guess the other code.
 - The beacon's frequency is configurable between 1 - 999. It also has a base code (synd or cent) which prevents ERTs from guessing into the syndicate beacon network.
 - You can toggle the beacon off, while still having the HUD on.
 - The HUD is toggelable.
 - The beacon only transmits while a suit is being worn.
 - The colour of each beacon is customisable in the code. (Syndicates always use their dark red, ERT use the colour of the loadout)
 - The marker beacon updates when you, or the thing you are tracking moves.
 - Implemented in code pretty simply so you can very easily set this up for other items too.
 - The markers have smooth transitions when turning.
 - Traitor bought hardsuits do not come with a syndicate code, however if you have some friends you can coordinate and share codes to add yourself to the tracking network.

Also yes, If one of your nukies is unrobust enough to lose their suit, the new owner will be able to see the location of the other suits in the team, however if that situation occurs you can disable your suit's tracking beacon.

Screenshots:
image
image
image

## Why It's Good For The Game

Really helps you to find your teammates, a pretty cool and nifty feature to have especially for medics.
Improves coordination at times when asking for locations is difficult.

## Changelog
:cl:
add: Adds in team tracking beacons for nukies and ERTs
add: Nukies and ERTs can now see the location of their teammates.
/:cl:

### CODE REVIEW

**Comment by bloons3** [comment]
Will this be ported to syndicate (Nuke Op) cyborgs?

Their pinpointer points to the nearest living syndicate.

File: code/modules/antagonists/nukeop/nukeop.dm
Comment (utilbird) [inline code/modules/antagonists/nukeop/nukeop.dm]:
```suggestion
	if(nuke_team?.team_frequency)
```

File: code/modules/clothing/spacesuits/hardsuit.dm
Comment (utilbird) [inline code/modules/clothing/spacesuits/hardsuit.dm:357]: oh god, how'd this go unnoticed for so long lmao

**Review by utilbird — COMMENTED** [review]
This is really cool, good job. I'd love to see this implemented with tools like the disk/med pointer.

File: code/modules/clothing/spacesuits/hardsuit.dm
Comment (PowerfulBacon) [reply code/modules/clothing/spacesuits/hardsuit.dm:357]: Combat combat combat combat combat combat

Comment (PowerfulBacon) [reply code/modules/clothing/spacesuits/hardsuit.dm:357]: Also this doesn't fix it, I need to add something else too

**Comment by heepox** [comment]
very nice.

File: code/datums/components/team_monitor.dm
Comment (PowerfulBacon) [inline code/datums/components/team_monitor.dm]:
```suggestion
```

**Comment by PowerfulBacon** [comment]
When the thing that is being tracked for movement loc changes, the component needs to start tracking that instead

**Comment by PowerfulBacon** [comment]
Also the really overpowered items should be locked to the ninja like before

**Comment by ItsTisme** [comment]
Very based, only suggestion maby have it as a implant or smth for nukies so that they can still track eatchother without wearing the suit for stealth ops

**Comment by PowerfulBacon** [comment]
Ok the monitor will now update every second, this will smooth out situations in which someone is inside another object which causes the moved signal not to get called. Should be ready to go at this point.

File: code/game/machinery/syndicatebomb.dm
Comment (park66665) [inline code/game/machinery/syndicatebomb.dm:53]: Every component of a datum is qdeleted at `/datum/proc/Destroy`. You don't have to explicitly call qdel on individual components.

File: code/game/machinery/syndicatebomb.dm
Comment (park66665) [inline code/game/machinery/syndicatebomb.dm:83]: ditto.

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm]: On `Destroy`, you should call parent proc *last*, not first.

File: code/modules/antagonists/ert/ert.dm
Comment (park66665) [inline code/modules/antagonists/ert/ert.dm:187]: ERT wants a better way to actually memorize tracking frequency,

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm]: `get_active_trackers` is a *really* bad name for this proc. Rename it to something like `get_matching_beacons`

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm]:
```suggestion
	if(_global)
		team_frequency = "[_frequency_key]-GLOB"
	else if(_frequency)
		team_frequency = "[_frequency_key][_frequency]"
	else
		team_frequency = null
```

File: code/modules/antagonists/nukeop/equipment/nuclearbomb.dm
Comment (park66665) [inline code/modules/antagonists/nukeop/equipment/nuclearbomb.dm]:
```suggestion
		AddComponent(/datum/component/tracking_beacon, "synd", null, null, TRUE, "#ebeca1", TRUE, TRUE)
```

File: code/game/machinery/syndicatebomb.dm
Comment (park66665) [inline code/game/machinery/syndicatebomb.dm]:
```suggestion
	AddComponent(/datum/component/tracking_beacon, "synd", null, null, TRUE, "#ff2b2b", TRUE, TRUE)
```

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm]:
```suggestion
	var/new_freq = input(user, "Enter a new frequency (1 - 999):", "Frequency Change", 1) as num|null
	if(!new_freq)
		to_chat(user, "<span class='warning'>Invalid frequency. Encrypted tracking HUD disabled.</span>")
		return
	if(new_freq < 1 || new_freq > 999)
		to_chat(user, "<span class='warning'>Frequency is out of range. Must be between 1 and 999.</span>")
		return
```

`team_frequency = null` lines here are holes in the entire control flow. For example, the proc below (`set_frequency`) assumes that non-null `team_frequency` equals functioning monitor and vice versa, but this lines null `team_frequency` without disabling the monitor completely, making the monitor stuck in a limbo state where things are prone to break.

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm]:
```suggestion
	if(!new_freq)
		to_chat(user, "<span class='warning'>Invalid frequency. Encrypted tracking beacon disabled.</span>")
		return
	if(new_freq < 1 || new_freq > 999)
		to_chat(user, "<span class='warning'>Frequency is out of range. Must be between 1 and 999.</span>")
		return
```

Mostly ditto.

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm]:
```suggestion
	updating = target
	//Our hud is disabled
	if(!hud_visible)
		return
	//Start processing to update in weird situations
	START_PROCESSING(SSprocessing, src)
	//Register parent signal
	RegisterSignal(target, COMSIG_MOVABLE_MOVED, .proc/parent_moved)
```

`hud_visible` disables hud, yet `update_all_directions` calls that'll be caused by `START_PROCESSING` and `RegisterSignal` will update & create arrows *without* checking `hud_visible`. Please consider the above suggestion, or add checks at `update_all_directions`

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm]:
```suggestion
			updating.hud_used.team_finder_arrows -= screen
			qdel(screen)
			tracking[beacon] = null
			//Update their hud
			updating.hud_used.show_hud(updating.hud_used.hud_version, updating)
```

Redundant checks; `updating.hud_used` is already checked at the beginning of this proc.

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm:380]: ~~FYI, `COMSIG_MOVABLE_MOVED` doesn't get sent to an item held by a mob that moved. Your disky therefore will barely receive `COMSIG_MOVABLE_MOVED`.~~

**Review by park66665 — COMMENTED** [review]
I took a cursory look at your code. I might review again (some changes are not yet looked at, too) after my concerns are addressed.

File: code/game/machinery/syndicatebomb.dm
Comment (PowerfulBacon) [reply code/game/machinery/syndicatebomb.dm:53]: This assumes that the explosion will actually destroy the bomb.

File: code/game/machinery/syndicatebomb.dm
Comment (PowerfulBacon) [reply code/game/machinery/syndicatebomb.dm:83]: This assumes that the explosion will actually destroy the bomb.

File: code/game/machinery/syndicatebomb.dm
Comment (park66665) [reply code/game/machinery/syndicatebomb.dm:53]: Your comment doesn't make sense. But my comment doesn't make sense here too. i'll mark this resolved.

File: code/game/machinery/syndicatebomb.dm
Comment (park66665) [reply code/game/machinery/syndicatebomb.dm:83]: I have no idea what you meant. Are you removing the beacon when a bomb *tries* to explode?

File: code/datums/components/team_monitor.dm
Comment (PowerfulBacon) [reply code/datums/components/team_monitor.dm:380]: Found this out in testing and getting around it is very difficult, so there is a keyframe update every 1 second which provides almost realtime tracking.

File: code/game/machinery/syndicatebomb.dm
Comment (PowerfulBacon) [reply code/game/machinery/syndicatebomb.dm:83]: Yes, or when a bomb is disarmed

File: code/modules/antagonists/ert/ert.dm
Comment (PowerfulBacon) [reply code/modules/antagonists/ert/ert.dm:187]: ill add it to their notes too

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm:124]: I suggest a more thorough cleaning. You also have to clean `attached_beacon`.

```suggestion
	if(updating)
		hide_hud(updating)
	if(attached_beacon)
		if(attached_beacon.attached_monitor == src)
			attached_beacon.attached_monitor = null
		attached_beacon = null
```

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm:409]: You should clean `attached_monitor` and its `attached_beacon` accordingly.

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm]:
```suggestion
	if(frequency)
```

`frequency` won't be 0.

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm:383]: I think you might be able to utilize `COMSIG_ITEM_PICKUP`.

File: code/datums/components/team_monitor.dm
Comment (park66665) [inline code/datums/components/team_monitor.dm:277]: This proc is not used. Beacons have an item action to `change_frequency`, but not monitors. You made beacon's `change_frequency` to change attached monitor's frequency with `set_frequency`, so you have two options:

1. Remove this proc. There currently isn't a monitor without attached beacon anyway.
2. Make separate item action to change monitor's frequency, ideally with a more complex input system like enabling the choice to (beacon) GLOB frequency and making frequencies synchronize with each other on demand.

**Review by park66665 — COMMENTED** [review]
I took another look at the changes. I don't think there will be another long review comment before merge.

File: code/datums/components/team_monitor.dm
Comment (PowerfulBacon) [reply code/datums/components/team_monitor.dm:277]: This proc exists for items that may be implemented in the future that contain a monitor but not a beacon. Without it, it means that it would have to be recoded.

File: code/datums/components/team_monitor.dm
Comment (PowerfulBacon) [reply code/datums/components/team_monitor.dm:383]: Equip is better since it also handles putting items on. In this case its about hardsuits being equipped since while a hardsuit isn't worn the beacon is disabled.

File: code/datums/components/team_monitor.dm
Comment (PowerfulBacon) [reply code/datums/components/team_monitor.dm:124]: Beacons and HUDs aren't necessarily destroyed together

Comment (PowerfulBacon) [reply code/datums/components/team_monitor.dm:124]: ill have a look at better cleaning

Comment (PowerfulBacon) [reply code/datums/components/team_monitor.dm:124]: Ah yea attached_beacon needs to be set to null

File: code/datums/components/team_monitor.dm
Comment (park66665) [reply code/datums/components/team_monitor.dm:383]: I was thinking about `always_update` ones, but it is your call.

File: code/datums/components/team_monitor.dm
Comment (park66665) [reply code/datums/components/team_monitor.dm:277]: I wasn't 100% content with the current state of that code; it looks as if it was a leftover code. For example, monitor's `change_frequency` doesn't set frequency of `attached_beacon`. Obviously, you decide. If *you* are content with the current state of code, then I'll gloss over this bit. However, I'd like you to add a comment about it if this proc really is intended for beacon-less monitors.

Comment (PowerfulBacon) [reply code/datums/components/team_monitor.dm:277]: I guess checking for an attached_beacon and changing that too would make sense for consistency sake

**Review by park66665 — APPROVED** [review]
All clear. This PR is good to go.

**Pull request merged**
